<a href="https://colab.research.google.com/github/Cjeyms/CCRNFLRL_EXERCISES_COM221ML/blob/main/Exercise2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercise 2

In [275]:
from math import sqrt, log
import pandas as pd

## Step 1: Enter your data

Add a new row after each pull: Pull number, Machine played ("A", "B", "C"), and Reward earned.

In [276]:
data = pd.DataFrame({
    "Pull": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
             27, 28, 29, 30, 31,32, 33, 34, 35,36, 37, 38,39,40, 41,42, 43, 44, 45, 46,47, 48, 49,50],          # Pull number
    "Machine": ["A", "B", "C", "A", "B", "C", "C", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A",
                "A", "A", "A", "A", "A", "A", "A", "A","A","A", "B", "A", "A", "A",
                "A","A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "B","B"], # Which machine was played
    "Reward": [0, 0, 0, 0, 0, 0, 0, 20, 0, 0, 0, 0, 5, 0, 1, 1, 2, 0, 2, 0, 0, 0, 0, 0, 0, 1,5,0, 0, 0, 0,0,0,0,5, 1,0,0, 0, 0,5,0, 0, 0, 0, 0,0, 0,5,0]         # Reward received from that machine
})

# Step 2: Calculate UCB

In [277]:
def calculate_ucb(df):
    results = []
    total_pulls = len(df)
    machines = sorted(df["Machine"].unique())

    for machine in machines:
        machine_data = df[df["Machine"] == machine]
        n_j = len(machine_data)
        avg_reward = machine_data["Reward"].mean()
        exploration_bonus = sqrt((2 * log(total_pulls)) / n_j) if n_j > 0 else float("inf")
        ucb_score = avg_reward + exploration_bonus
        results.append({
            "Machine": machine,
            "Pulls": n_j,
            "Avg Reward": round(avg_reward, 2),
            "Exploration Bonus": round(exploration_bonus, 4),
            "UCB Score": round(ucb_score, 4)
        })
    return pd.DataFrame(results).sort_values(by="UCB Score", ascending=False)


# Step 3: Calculate and display


In [278]:
ucb_table = calculate_ucb(data)
print("📊 UCB Results Table")
print(ucb_table)

📊 UCB Results Table
  Machine  Pulls  Avg Reward  Exploration Bonus  UCB Score
1       B      5        1.00             1.2509     2.2509
2       C      3        0.00             1.6149     1.6149
0       A     42        1.14             0.4316     1.5745


# Step 4: Next machine suggestion

In [279]:
best_machine = ucb_table.iloc[0]["Machine"]
print(f"\n🎯 Recommended next machine to play: {best_machine}")


🎯 Recommended next machine to play: B


# Step 4: REFLECTION


 Based on the results, Slot A had the highest average reward at 1.14, while Slot B had 1.00 and Slot C had 0.00. The code ended up pulling Slot A the most (42 times) because it consistently gave higher rewards compared to the other slots. This behavior shows that the code/system able to recognize which option provided the best payoff. At first, it explored different slots, but over time it shifted to exploitation—focusing more on Slot A to maximize the total reward.